File: py_src/paperjam/_convert.py
"""Format conversion functions."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import os

from paperjam import _paperjam


def convert(
    input_path: str | os.PathLike[str],
    output_path: str | os.PathLike[str],
) -> dict:
    """Convert a file from one format to another.

    Formats are auto-detected from file extensions.

    Returns:
        A dict with keys: from_format, to_format, content_blocks, tables, images.
    """
    return dict(_paperjam.convert_file(str(input_path), str(output_path)))


def convert_bytes(
    data: bytes,
    *,
    from_format: str,
    to_format: str,
) -> bytes:
    """Convert in-memory bytes from one format to another.

    Args:
        data: Source document bytes.
        from_format: Source format (e.g., 'pdf', 'docx', 'html').
        to_format: Target format (e.g., 'pdf', 'markdown', 'html').

    Returns:
        Converted document bytes.
    """
    return bytes(_paperjam.convert_bytes(data, from_format, to_format))


def detect_format(path: str | os.PathLike[str]) -> str:
    """Detect the document format from a file path.

    Returns:
        Format string (e.g., 'pdf', 'docx', 'html', 'epub', 'unknown').
    """
    return str(_paperjam.detect_format(str(path)))

File: crates/paperjam-convert/src/convert.rs
use std::path::Path;

use paperjam_model::format::DocumentFormat;

use crate::detect::detect_format;
use crate::error::ConvertError;
use crate::extract;
use crate::generate;

/// Summary of a completed conversion.
pub struct ConvertReport {
    pub from_format: DocumentFormat,
    pub to_format: DocumentFormat,
    pub content_blocks: usize,
    pub tables: usize,
    pub images: usize,
}

/// Convert a file from one format to another, reading from `input_path` and
/// writing to `output_path`.  The source and target formats are inferred from
/// file extensions (with magic-byte fallback for the input).
pub fn convert(input_path: &Path, output_path: &Path) -> Result<ConvertReport, ConvertError> {
    let input_bytes = std::fs::read(input_path)?;
    let from_format = detect_format(input_path);
    let to_format = DocumentFormat::detect(output_path);

    if from_format == DocumentFormat::Unknown {
        return Err(ConvertError::unsupported(from_format));
    }
    if to_format == DocumentFormat::Unknown {
        return Err(ConvertError::unsupported(to_format));
    }

    let intermediate = extract::extract(&input_bytes, from_format)?;
    let output_bytes = generate::generate(&intermediate, to_format)?;

    std::fs::write(output_path, &output_bytes)?;

    Ok(ConvertReport {
        from_format,
        to_format,
        content_blocks: intermediate.blocks.len(),
        tables: intermediate.tables.len(),
        images: intermediate.images.len(),
    })
}

/// Convert in-memory bytes from one format to another.
pub fn convert_bytes(
    input: &[u8],
    from_format: DocumentFormat,
    to_format: DocumentFormat,
) -> Result<Vec<u8>, ConvertError> {
    let intermediate = extract::extract(input, from_format)?;
    generate::generate(&intermediate, to_format)
}
